### ISSUE TITLE
Allow cancelling execution on KeyboardInterrupt

### ISSUE THREAD

**Issue by akdor1154** [pola-rs/polars#2687]
#### Describe your feature request

It'd be great if I could cancel a big computation with python's KeyboardInterrupt.

Happy to look into this myself as a first jump into the code - raising first here for visibility. (my perspective is from python interactive use, but assuming work for this is mainly on the rust side, as that's where your thread pool seems to be. could be completely off the mark is as I haven't dived in yet.)

**Comment by ritchie46** [comment]
This is hard because I don't want polars core to have anything to do with python.

Pyo3 docs: 

> [Ctrl-C doesn't do anything while my Rust code is executing!](https://pyo3.rs/v0.13.2/faq.html#ctrl-c-doesnt-do-anything-while-my-rust-code-is-executing)

This is because Ctrl-C raises a SIGINT signal, which is handled by the calling Python process by simply setting a flag to action upon later. This flag isn't checked while Rust code called from Python is executing, only once control returns to the Python interpreter.

> You can give the Python interpreter a chance to process the signal properly by calling Python::check_signals. It's good practice to call this function regularly if you have a long-running Rust function so that your users can cancel it.

**Comment by ghuls** [comment]
I think Python C extensions (like numpy) have the same problem, so I think it is something fundamental in python that it does not pass Ctrl+C to extension modules.

**Comment by zundertj** [comment]
Marking as a wont-fix because it seems unlikely we could take on something this fundamental, although I agree with that it would a great feature to have.

@akdor1154 : if you think you could resolve this yourself, I am sure that would be a very welcome contribution.

**Comment by bverhoeff** [comment]
Apparently it should be possible to propagate the ctrl-c to rust by starting a Rust thread in Python. I don't know polars well enough to integrate this. Furthermore, the API would change considerably so this might involve a separate/additional Polars workflow:

To make the `while True` loop in Python stop when the Rust function completes, you need a mechanism to check the state of the Rust function. This can be done by having the Rust function signal its completion back to Python. Here’s how you can achieve this:

### Rust Code:

```rust
use pyo3::prelude::*;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;

#[pyclass]
struct RustWorker {
    completed: Arc<AtomicBool>, // To track if the work is completed
    interrupted: Arc<AtomicBool>, // To handle interruption
}

#[pymethods]
impl RustWorker {
    #[new]
    fn new() -> Self {
        RustWorker {
            completed: Arc::new(AtomicBool::new(false)),
            interrupted: Arc::new(AtomicBool::new(false)),
        }
    }

    fn start_work(&self) -> PyResult<()> {
        let completed = self.completed.clone();
        let interrupted = self.interrupted.clone();

        thread::spawn(move || {
            for i in 0..10 {
                if interrupted.load(Ordering::SeqCst) {
                    eprintln!("Work interrupted!");
                    break;
                }
                eprintln!("Working... {}", i);
                std::thread::sleep(std::time::Duration::from_secs(1));
            }

            completed.store(true, Ordering::SeqCst); // Signal completion
            eprintln!("Work completed!");
        });

        Ok(())
    }

    fn is_completed(&self) -> PyResult<bool> {
        Ok(self.completed.load(Ordering::SeqCst)) // Return completion status
    }

    fn interrupt(&self) -> PyResult<()> {
        self.interrupted.store(true, Ordering::SeqCst);
        Ok(())
    }
}
```

### Python code:

```python
import rust_worker
import time

def main():
    worker = rust_worker.RustWorker()

    try:
        print("Starting work...")
        worker.start_work()

        while True:
            if worker.is_completed():  # Check if the Rust function is completed
                print("Work completed!")
                break
            time.sleep(0.5)  # Polling interval
    except KeyboardInterrupt:
        print("Interrupt received! Stopping work...")
        worker.interrupt()
    finally:
        print("Exiting main loop.")

if __name__ == "__main__":
    main()
```

**Comment by ritchie46** [comment]
For people running Polars in the terminal and just want to kill the process: `ctrl+/` is your friend.

**Comment by bverhoeff** [comment]
But when running a python script that imports polars functionality and running a long polars script that way `ctrl+/` or `ctrl+c` doesn't work.

**Comment by ritchie46** [comment]
`ctrl+/` works for me. Did you try it? I believe only `ctrl+c` doesn't work as that is catched by Python.

**Comment by adamreeve** [comment]
I was looking into this recently and for the record, it looks like Ctrl+C handling was added in #20826 (released in 1.21.0) and #20961 (released in 1.22.0)
